File: 46-permutations/46-permutations.cpp
class Solution {
public:
    
    // Backtracking
    void permutation(vector<int> &nums,vector<vector<int>> &result, int i,int n){
        if(i==n){
            result.push_back(nums);
            return ;
        }

        for(int j=i;j<=n;j++){
            swap( nums[i],nums[j]);
            permutation(nums,result,i+1,n);
            swap( nums[i],nums[j]);
        }
    }
    
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        
        permutation(nums,result,0,nums.size()-1);
        
        return result;
    }

};

File: 46-permutations/46-permutations.js
function permute(nums) {
  const result = [];
  const visited = new Set();
  dfs(nums, result, [], visited);
  return result;
};

/*
  - permutation is used to keep track of current permutation
  - seen set is used to avoid using the same index twice
  - reuslt array is used to store the permutations
*/
function dfs(nums, result, permutation, visited) {
  if(permutation.length === nums.length){
    result.push([...permutation]);
    return;
  }
  
  for(let i = 0 ; i < nums.length ; i++){
    if(visited.has(i)) continue;
    
    // include the number at end of permutation
    permutation.push(nums[i]);
    visited.add(i);
    
    // continue dfs
    dfs(nums, result, permutation, visited);
    
    visited.delete(i);
    permutation.pop();
  }
}